File: EightQueens.java
import java.util.Random;

public class EightQueens {

    private static final int SIZE = 8;
    private static final int MAX_RESTARTS = 1000;
    private int[] board; // This represents the positions of queens in each row

    public EightQueens() {
        board = new int[SIZE];
    }

    // Randomly initialize the board
    private void initializeBoard() {
        Random rand = new Random();
        for (int i = 0; i < SIZE; i++) {
            board[i] = rand.nextInt(SIZE);
        }
    }

    // Calculate the number of conflicts (pairs of queens attacking each other)
    private int calculateConflicts(int[] state) {
        int conflicts = 0;
        for (int i = 0; i < SIZE; i++) {
            for (int j = i + 1; j < SIZE; j++) {
                if (state[i] == state[j] || Math.abs(state[i] - state[j]) == Math.abs(i - j)) {
                    conflicts++;
                }
            }
        }
        return conflicts;
    }

    // Perform the Hill Climbing algorithm with a random restart
    private int[] hillClimbWithRandomRestart() {
        int restartCount = 0;
        int[] currentState = new int[SIZE];
        int[] nextState;
        int currentConflicts;
        int nextConflicts;

        while (restartCount < MAX_RESTARTS) {
            initializeBoard();
            currentState = board.clone();
            currentConflicts = calculateConflicts(currentState);

            boolean localOptimumReached = false;
            while (!localOptimumReached) {
                nextState = currentState.clone();
                int bestConflicts = currentConflicts;

                // Explore the neighborhood
                for (int col = 0; col < SIZE; col++) {
                    for (int row = 0; row < SIZE; row++) {
                        if (currentState[col] == row) {
                            continue;
                        }
                        nextState[col] = row;
                        nextConflicts = calculateConflicts(nextState);

                        // If the new state has fewer conflicts, move to that state
                        if (nextConflicts < bestConflicts) {
                            bestConflicts = nextConflicts;
                            currentState[col] = row;
                        }

                        // Reset nextState to currentState for the next iteration
                        nextState[col] = currentState[col];
                    }
                }

                if (bestConflicts == currentConflicts) {
                    // No better neighbor found; local optimum reached
                    localOptimumReached = true;
                } else {
                    currentConflicts = bestConflicts;
                }

                // Check if a solution has been found
                if (currentConflicts == 0) {
                    return currentState;
                }
            }
            restartCount++;
        }
        return null; // Return null if no solution is found within MAX_RESTARTS
    }

    // Print the board
    private void printBoard(int[] state) {
        if (state == null) {
            System.out.println("No solution found.");
            return;
        }

        for (int row = 0; row < SIZE; row++) {
            for (int col = 0; col < SIZE; col++) {
                if (state[col] == row) {
                    System.out.print("Q ");
                } else {
                    System.out.print(". ");
                }
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        EightQueens eightQueens = new EightQueens();
        int[] solution = eightQueens.hillClimbWithRandomRestart();
        eightQueens.printBoard(solution);
    }
}

File: EightQueens.py
import random

def random_board(n=8):
    """Generate a random board for n-queens problem."""
    return [random.randint(0, n-1) for _ in range(n)]

def count_attacks(board):
    """Count the number of attacks on the board."""
    attacks = 0
    n = len(board)
    for i in range(n):
        for j in range(i + 1, n):
            if board[i] == board[j] or abs(board[i] - board[j]) == abs(i - j):
                attacks += 1
    return attacks

def get_neighbors(board):
    """Generate all possible neighboring states."""
    neighbors = []
    n = len(board)
    for i in range(n):
        for j in range(n):
            if board[i] != j:
                neighbor = board[:]
                neighbor[i] = j
                neighbors.append(neighbor)
    return neighbors

def hill_climbing(board):
    """Perform hill climbing to solve the n-queens problem."""
    current_board = board
    current_attacks = count_attacks(current_board)

    while True:
        neighbors = get_neighbors(current_board)
        next_board = min(neighbors, key=count_attacks)
        next_attacks = count_attacks(next_board)

        if next_attacks >= current_attacks:
            return current_board

        current_board = next_board
        current_attacks = next_attacks

def hill_climbing_with_random_restarts(n=8, max_restarts=1000):
    """Solve n-queens problem with hill climbing and random restarts."""
    for _ in range(max_restarts):
        board = random_board(n)
        solution = hill_climbing(board)
        if count_attacks(solution) == 0:
            return solution
    return None

# Solve the 8-queens problem
solution = hill_climbing_with_random_restarts()
if solution:
    print("Solution found:", solution)
else:
    print("No solution found after max restarts.")